File: backend/src/Fix.Application/UseCases/Organizations/Mappers/OrganizationMapper.cs
using Fix.Application.Abstractions.Authentication;
using Fix.Application.Organizations.Dtos;
using Fix.Domain.AggregateRoots.Organizations;

namespace Fix.Application.Organizations.Mappers;

internal static class OrganizationMapper
{
    public static OrganizationDto ToDto(this Organization organization) =>
        new(organization.Id, organization.Name.Value, organization.Slug.Value);

    public static OrganizationSetupDto ToSetupDto(this Organization organization) => new(
        organization.Id,
        organization.Name.Value,
        organization.Slug.Value,
        new CompanyProfileDto(
            organization.Profile.CorporateName,
            organization.Profile.TaxId,
            organization.Profile.Headquarters,
            organization.Profile.Group,
            organization.Profile.Sector,
            organization.Profile.CropYearStartMonth,
            organization.Profile.ActiveCrop),
        new IndustrialProfileDto(
            organization.Industrial.MillingCapacity,
            organization.Industrial.MixMinPct,
            organization.Industrial.MixMaxPct,
            organization.Industrial.MixGuidancePct),
        new BudgetDto(
            organization.Budget.CashCost,
            organization.Budget.EconomicFloor,
            organization.Budget.EquivalentPrice,
            organization.Budget.TargetMarginPct),
        new FinancialsDto(
            organization.Financials.Cash,
            organization.Financials.CreditLines,
            organization.Financials.MonthlyFixedCost,
            organization.Financials.NetDebt,
            organization.Financials.Ebitda,
            organization.Financials.UsdDebt,
            organization.Financials.ReferenceDate,
            organization.Financials.Leverage),
        organization.Commodities
            .OrderBy(c => c.Commodity)
            .Select(c => new CommodityDto(c.Id, c.Commodity, c.Capacity, c.Unit, c.PriceReference, c.Currency, c.Sells))
            .ToList());

    public static IReadOnlyList<MemberDto> ToMemberDtos(
        this Organization organization,
        IReadOnlyDictionary<Guid, UserInfo> users,
        IReadOnlyDictionary<Guid, string> ruleCodes) =>
        organization.Members
            .Select(member =>
            {
                users.TryGetValue(member.UserId, out var user);

                var rules = organization.Rules
                    .Where(r => r.MemberId == member.Id)
                    .Select(r => ruleCodes.GetValueOrDefault(r.RuleId, r.RuleId.ToString()))
                    .ToList();

                var groups = organization.Groups
                    .Where(g => g.Members.Any(gm => gm.MemberId == member.Id))
                    .Select(g => g.Name.Value)
                    .ToList();

                return new MemberDto(
                    member.Id,
                    member.UserId,
                    user?.Email ?? string.Empty,
                    user?.FullName ?? string.Empty,
                    member.Desk,
                    rules,
                    groups);
            })
            .ToList();

    public static IReadOnlyList<GroupDto> ToGroupDtos(
        this Organization organization,
        IReadOnlyDictionary<Guid, string> ruleCodes) =>
        organization.Groups
            .Select(group => new GroupDto(
                group.Id,
                group.Name.Value,
                group.IsDefault,
                organization.Rules
                    .Where(r => r.GroupId == group.Id)
                    .Select(r => ruleCodes.GetValueOrDefault(r.RuleId, r.RuleId.ToString()))
                    .ToList(),
                group.Members.Select(m => m.MemberId).ToList()))
            .ToList();
}

File: frontend/bff/src/application/organizations/mappers/organization.mapper.ts
import type { CommodityDto } from '../dtos/commodity.dto.js';
import type { GroupDto } from '../dtos/group.dto.js';
import type { MemberDto } from '../dtos/member.dto.js';
import type { OrganizationSetupDto } from '../dtos/organization-setup.dto.js';
import type { OrganizationDto } from '../dtos/organization.dto.js';
import type { CommodityResponse } from '../responses/commodity.response.js';
import type { GroupResponse } from '../responses/group.response.js';
import type { MemberResponse } from '../responses/member.response.js';
import type { OrganizationSetupResponse } from '../responses/organization-setup.response.js';
import type { OrganizationResponse } from '../responses/organization.response.js';

export const toOrganizationResponse = (dto: OrganizationDto): OrganizationResponse => ({
  id: dto.id,
  name: dto.name,
  slug: dto.slug,
});

export const toCommodityResponse = (dto: CommodityDto): CommodityResponse => ({ ...dto });

export const toOrganizationSetupResponse = (dto: OrganizationSetupDto): OrganizationSetupResponse => ({
  id: dto.id,
  name: dto.name,
  slug: dto.slug,
  profile: { ...dto.profile },
  industrial: { ...dto.industrial },
  budget: { ...dto.budget },
  financials: { ...dto.financials },
  commodities: dto.commodities.map(toCommodityResponse),
});

export const toMemberResponse = (dto: MemberDto): MemberResponse => ({ ...dto, rules: [...dto.rules], groups: [...dto.groups] });

export const toGroupResponse = (dto: GroupDto): GroupResponse => ({ ...dto, rules: [...dto.rules], memberIds: [...dto.memberIds] });
